File: src/main/java/com/velha/GameLogic.java
package com.velha;

public class GameLogic {

    private static final char[] board = new char[9];
    private static char currentPlayer = 'X';
    private static boolean gameOver = false;
    private static int timeLimit = 10;
    private static String winnerMessage = "";

    private static final int[][] WIN_COMBOS = {
            {0, 1, 2}, {3, 4, 5}, {6, 7, 8},
            {0, 3, 6}, {1, 4, 7}, {2, 5, 8},
            {0, 4, 8}, {2, 4, 6}
    };

    public static void resetBoard() {
        for (int i = 0; i < 9; i++) {
            board[i] = '\0';
        }
        currentPlayer = 'X';
        gameOver = false;
        winnerMessage = "";
    }

    public static boolean makeMove(int position) {
        if (position < 0 || position > 8) return false;
        if (board[position] != '\0') return false;
        if (gameOver) return false;

        board[position] = currentPlayer;
        return true;
    }

    public static void switchPlayer() {
        if (currentPlayer == 'X') {
            currentPlayer = 'O';
        } else {
            currentPlayer = 'X';
        }
    }

    public static char checkWinner() {
        for (int i = 0; i < WIN_COMBOS.length; i++) {
            int a = WIN_COMBOS[i][0];
            int b = WIN_COMBOS[i][1];
            int c = WIN_COMBOS[i][2];

            if (board[a] != '\0' && board[a] == board[b] && board[b] == board[c]) {
                gameOver = true;
                winnerMessage = "Jogador " + board[a] + " venceu!";
                return board[a];
            }
        }
        return '\0';
    }

    public static boolean checkDraw() {
        for (int i = 0; i < 9; i++) {
            if (board[i] == '\0') return false;
        }
        gameOver = true;
        winnerMessage = "Empate!";
        return true;
    }

    public static boolean isGameOver() {
        return gameOver;
    }

    public static char getCurrentPlayer() {
        return currentPlayer;
    }

    public static String getWinnerMessage() {
        return winnerMessage;
    }

    public static int getTimeLimit() {
        return timeLimit;
    }

    public static void setTimeLimit(int seconds) {
        timeLimit = seconds;
    }

    public static void setGameOver(boolean over) {
        gameOver = over;
    }

    public static void setWinnerMessage(String msg) {
        winnerMessage = msg;
    }
}

File: src/VelhaGame.App/GameLogic.fs
module VelhaGame.GameLogic

open VelhaGame.Types

/// Combinações vencedoras (índices do tabuleiro)
let winCombos = [
    [0; 1; 2]
    [3; 4; 5]
    [6; 7; 8]
    [0; 3; 6]
    [1; 4; 7]
    [2; 5; 8]
    [0; 4; 8]
    [2; 4; 6]
]

/// Cria um tabuleiro vazio (9 células Empty)
let emptyBoard : Board = List.replicate 9 Empty

/// Verifica se uma lista de 3 células é uma vitória para um jogador
let allSame (symbol: Cell) (cells: Cell list) =
    cells |> List.forall (fun c -> c = symbol)

/// Verifica o resultado do jogo dado um tabuleiro
let checkResult (board: Board) : GameResult =
    let getCells (indices: int list) : Cell list =
        indices |> List.map (fun i -> board.[i])

    let winFor (player: Player) : bool =
        let symbol = player.Symbol
        winCombos
        |> List.exists (fun combo -> allSame symbol (getCells combo))

    if winFor PlayerX then Win PlayerX
    elif winFor PlayerO then Win PlayerO
    elif board |> List.forall (fun c -> c <> Empty) then Draw
    else Ongoing

/// Troca o jogador atual
let switchPlayer (player: Player) : Player =
    match player with
    | PlayerX -> PlayerO
    | PlayerO -> PlayerX

/// Tenta fazer uma jogada em uma posição. Retorna None se inválido.
let tryMakeMove (position: int) (board: Board) (player: Player) : Board option =
    if position < 0 || position > 8 then
        None
    else
        let cell = board.[position]
        match cell with
        | Empty ->
            let newBoard =
                board
                |> List.mapi (fun i c ->
                    if i = position then player.Symbol else c)
            Some newBoard
        | _ -> None

/// Verifica se o jogo acabou
let isGameOver (result: GameResult) : bool =
    result <> Ongoing

/// Gera a mensagem de vencedor
let winnerMessage (result: GameResult) : string =
    match result with
    | Win player -> sprintf "%s venceu!" player.DisplayName
    | Draw -> "Empate!"
    | Ongoing -> ""

/// Cria o estado inicial do menu
let initMenu () : Model = {
    Screen = Menu
    Board = emptyBoard
    CurrentPlayer = PlayerX
    Result = Ongoing
    WinnerMessage = ""
}

/// Cria um novo estado de jogo
let initGame () : Model = {
    Screen = Game
    Board = emptyBoard
    CurrentPlayer = PlayerX
    Result = Ongoing
    WinnerMessage = ""
}
